src/logicTemplates/dynamicfee.js

```diff
@@ -65,95 +65,112 @@ class DynamicFee {
+ * @returns {Object} object containing json of txnbuilder constructor arguments under "txn" and signed logicsig under "lsig"
+ */
+function signDynamicFee(contract, secretKey, genesisHash) {
+    let programOutputs = logic.readProgram(contract, undefined);
+    let ints = programOutputs[0];
+    let byteArrays = programOutputs[1];
+    let keys = nacl.keyPairFromSecretKey(secretKey);
+    let from = address.encode(keys.publicKey);
+    let to = address.encode(byteArrays[0]);
+    let fee = 0;
+    let amount = ints[2];
+    let closeRemainderTo = address.encode(byteArrays[1]);
+    let firstRound = ints[3];
+    let lastRound = ints[4];
+    let lease = new Uint8Array(byteArrays[2]);
+    let txn = {
+        "from": from,
+        "to": to,
+        "fee": fee,
+        "amount": amount,
+        "closeRemainderTo": closeRemainderTo,
+        "firstRound": firstRound,
+        "lastRound": lastRound,
+        "genesisHash": genesisHash,
+        "type": "pay",
+        "lease": lease
+    };
+
+    let lsig = new logicSig.LogicSig(contract, undefined);
+    lsig.sign(secretKey);
+    return {"txn": txn, "lsig": lsig};
 }
 
 /**
- * GetDynamicFeeTransactions creates and signs the secondary dynamic fee transaction, updates
+ * getDynamicFeeTransactions creates and signs the secondary dynamic fee transaction, updates
  * transaction fields, and signs as the fee payer; it returns both
  * transactions as bytes suitable for sendRaw.
  * Parameters:
- * @param {Transaction} txn - main transaction from payer
- * @param {LogicSig} lsig - the signed logic received from the payer
+ * @param {dict} txn - main transaction from payer's signDynamicFee output (a dict of constructor arguments, NOT a transaction.Transaction)
+ * @param {LogicSig} lsig - the signed logic received from the payer's signDynamicFee output
  * @param {Uint8Array} privateKey - the private key for the account that pays the fee
  * @param {int} fee - fee per byte for both transactions
- * @param {int} firstValid - first protocol round on which both transactions will be valid
- * @param {int} lastValid - last protocol round on which both transactions will be valid
  *
+ * @throws on invalid lsig
  */
-function getDynamicFeeTransactions (txn, lsig, privateKey, fee, firstValid, lastValid) {
-    txn.firstRound = firstValid;
-    txn.lastRound = lastValid;
+function getDynamicFeeTransactions (txn, lsig, privateKey, fee) {
+    if (!lsig.verify(address.decode(txn.from).publicKey)) {
+        throw new Error("invalid signature");
+    }
+
     txn.fee = fee;
-    txn.fee *= txn.estimateSize();
     if (txn.fee < transaction.ALGORAND_MIN_TX_FEE) {
         txn.fee = transaction.ALGORAND_MIN_TX_FEE
     }
 
     let keys = nacl.keyPairFromSecretKey(privateKey);
-    let address = address.encode(keys.publicKey);
+    let from = address.encode(keys.publicKey);
+
+    // must remove lease and re-add using addLease so that fee calculation will match other SDKs
```

Comment:
This has been discussed, but maybe we should change this in all the SDKs to make for more accurate fee calculations, since if doesn't include the lease the fee is substantially less than it should be.